src/middlewares/createTelemetryMiddleware.ts

```diff
@@ -1,11 +1,70 @@
-import { Middleware } from '../types';
+import { InstantSearch, Middleware, Widget } from '../types';
+import { Index } from '../widgets/index/index';
 
 type TelemetryWidget = {
   type: string;
   params: string[];
   officialWidget: boolean;
 };
 
+type TelemetryPayload = {
+  widgets: TelemetryWidget[];
+};
+
+function extractPayload(
+  widgets: Array<Widget<{ renderState: any }>>,
+  instantSearchInstance: InstantSearch,
+  payload: TelemetryPayload
+) {
+  const parent = instantSearchInstance.mainIndex;
+
+  const initOptions = {
+    instantSearchInstance,
+    parent,
+    scopedResults: parent.getScopedResults(),
+    state: parent.getHelper()!.state,
+    helper: parent.getHelper()!,
+    createURL: parent.createURL,
+    uiState: instantSearchInstance._initialUiState,
+    renderState: instantSearchInstance.renderState,
+    templatesConfig: instantSearchInstance.templatesConfig,
+    searchMetadata: {
+      isSearchStalled: instantSearchInstance._isSearchStalled,
+    },
+  };
+
+  widgets.forEach(widget => {
+    let widgetParams = {};
+
+    if (widget.getWidgetRenderState) {
+      const renderState = widget.getWidgetRenderState(initOptions);
+
+      if (renderState && renderState.widgetParams) {
+        widgetParams = renderState.widgetParams;
+      }
+    }
+
+    // since we destructure in all widgets, the parameters with defaults are set to "undefined"
+    const params = Object.keys(widgetParams).filter(
+      key => widgetParams[key] !== undefined
+    );
+
+    payload.widgets.push({
+      type: widget.$$type || 'custom.widget',
+      params,
+      officialWidget: Boolean(widget.$$officialWidget),
+    });
+
+    if (widget.$$type === 'ais.index') {
+      extractPayload(
```

Comment:
This seems to be gathering payloads in a flat array. Won't we lose information about the structure of nested indices?

Reply:
yes, we lose the information for now, but flat information will be much easier to search; as well as indexName not being anonymous

Reply:
Ah right. I've forgotten about it being stored in Algolia. And I didn't think of indexName. Then it makes sense to me to have it this way.

Reply:
I see these options:

1. ignore non-connector parameters for now
2. pass the original widgetParams to makeWidget as $$widgetParams
3. pass each of the widget params to the connector (even though it does nothing)